geopandas/tools/sjoin.py

```diff
@@ -27,131 +31,76 @@ def sjoin(left_df, right_df, how='inner', op='intersects',
-        left_df, right_df = right_df, left_df
-
-    # insert the bounds in the rtree spatial index
-    right_df_bounds = right_df.geometry.apply(lambda x: x.bounds)
-    stream = ((i, b, None) for i, b in enumerate(right_df_bounds))
-    tree_idx = rtree.index.Index(stream)
-
-    idxmatch = (left_df.geometry.apply(lambda x: x.bounds)
-                .apply(lambda x: list(tree_idx.intersection(x))))
-    idxmatch = idxmatch[idxmatch.apply(len) > 0]
-
-    if idxmatch.shape[0] > 0:
-        # if output from join has overlapping geometries
-        r_idx = np.concatenate(idxmatch.values)
-        l_idx = np.concatenate([[i] * len(v) for i, v in idxmatch.iteritems()])
-
-        # Vectorize predicate operations
-        def find_intersects(a1, a2):
-            return a1.intersects(a2)
-
-        def find_contains(a1, a2):
-            return a1.contains(a2)
-
-        predicate_d = {'intersects': find_intersects,
-                       'contains': find_contains,
-                       'within': find_contains}
-
-        check_predicates = np.vectorize(predicate_d[op])
-
-        result = (
-                  pd.DataFrame(
-                      np.column_stack(
-                          [l_idx,
-                           r_idx,
-                           check_predicates(
-                               left_df.geometry
-                               .apply(lambda x: prepared.prep(x))[l_idx],
-                               right_df[right_df.geometry.name]._geometry_array[r_idx])
-                           ]))
-                   )
-
-        result.columns = ['_key_left', '_key_right', 'match_bool']
-        result = (
-                  pd.DataFrame(result[result['match_bool']==1])
-                  .drop('match_bool', axis=1)
-                  )
-
+    indices = cysjoin(left_df.geometry._geometry_array.data,
+                      right_df.geometry._geometry_array.data,
+                      op)
+    n = len(indices)
+    left_indices = indices[:, 0]
+    right_indices = indices[:, 1]
+
+    if how == 'left':
+        missing = pd.Index(np.arange(len(left_df))).difference(pd.Index(left_indices))
+        if len(missing):
+            left_indices = np.concatenate([left_indices, missing])
+
+    if how == 'right':
+        missing = pd.Index(np.arange(len(right_df))).difference(pd.Index(right_indices))
+        if len(missing):
+            right_indices = np.concatenate([right_indices, missing])
+
+    n_left = len(left_indices)
+    n_right = len(right_indices)
+
+    left = left_df.take(left_indices)
+    right = right_df.take(right_indices)
+
+    if how in ('inner', 'left'):
+        del right[right._geometry_column_name]
+        index = left.index
+    else:
+        del left[left._geometry_column_name]
+        index = right.index
+
+    columns = {}
+    names = []
+    for name, series in left.iteritems():
+        if name in right.columns:
+            name = lsuffix + '_' + name
+        series.index = index[:n_left]
```

Comment:
ah, for `how='right'` this is not the case. But can't we put this one then in the if block below?